### PR TITLE
[aievec] Add aievec transform dialect extensions

### PR DESCRIPTION
**Pull request by jsetoain**
This patch wires extensions to the transform dialect for the AIEVec dialect, and adds a first `transform.structured` op to vectorize `linalg.generic` ops with a vector contraction as the innermost operation. This transform op can be used as a path to lower GEMM to matrix multiply operations in AIEv2. It also serves as a reference for further extensions to the Transform dialect.

### CODE REVIEW

File: include/aie/Dialect/AIEVec/TransformOps/AIEVecTransformOps.td [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+include "mlir/Dialect/Transform/IR/TransformTypes.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/IR/OpBase.td"
+
+def VectorizeContractionOp : Op<Transform_Dialect,
+                        "structured.vectorize_contraction", [
+                        FunctionalStyleTransformOpTrait,
+                        MemoryEffectsOpInterface, TransformEachOpTrait,
+                        TransformOpInterface]> {
+//                            ReportTrackingListenerFailuresOpTrait]> {
```
Comment (linay-xsj) [inline include/aie/Dialect/AIEVec/TransformOps/AIEVecTransformOps.td:24]: redundant

Comment (linay-xsj) [reply include/aie/Dialect/AIEVec/TransformOps/AIEVecTransformOps.td:24]: Several redundant code are needed to be removed in the comments.

File: test/dialect/AIEVec/Transform/invalid.mlir [hunk 2]
```diff
@@ -0,0 +1,182 @@
+// RUN: aie-opt %s -test-transform-dialect-interpreter -split-input-file -verify-diagnostics
```
Comment (linay-xsj) [inline test/dialect/AIEVec/Transform/invalid.mlir:1]: Do we need file check in this test case?

Comment (jsetoain) [reply test/dialect/AIEVec/Transform/invalid.mlir:1]: No, `-verify-diagnostics` is the option that consumes the `expected-error` directives.

File: include/aie/Dialect/AIEVec/TransformOps/AIEVecTransformOps.td [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+      Notice how the `affine_maps` describing the access patterns have been
+      split between tile accesses, in the `linalg.generic` op, and matrix
+      multiply accesses, in the `vector.contract` op.
+      }];
+
+  let arguments = (ins PDL_Operation:$target);
+  // TODO: Return all new ops generated neatly arranged
+  let results = (outs PDL_Operation:$transformed);
+//            $target attr-dict `:` functional-type($target, results)
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply include/aie/Dialect/AIEVec/TransformOps/AIEVecTransformOps.td:137]: See my answer above.

File: lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  auto opMemrefVecTy = MemRefType::get(newShape, opVecElemTy);
+  auto typeCastOp =
+      rewriter.create<vector::TypeCastOp>(loc, opMemrefVecTy, toMemRefOp);
+  // Missing a 'restrict' qualifier
+  auto toTensorOp = rewriter.create<bufferization::ToTensorOp>(
+      loc, RankedTensorType::get(newShape, opVecElemTy), typeCastOp);
+  // TODO: This won't work until we next update llvm, with the newly added
+  // TODO: properties. In the meantime, I'm not sure how to do this, or
+  // TODO: whether it matters at all.
+  // toTensorOp.setRestrict(true);
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp:64]: ditto

File: lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+  auto typeCastOp =
+      rewriter.create<vector::TypeCastOp>(loc, opMemrefScalTy, toMemRefVecTyOp);
+
+  // Missing a 'restrict' qualifier
+  auto toTensorOp = rewriter.create<bufferization::ToTensorOp>(
+      loc, RankedTensorType::get(scalShape, elemTy), typeCastOp);
+  // TODO: This won't work until we next update llvm, with the newly added
+  // TODO: properties. In the meantime, I'm not sure how to do this, or
+  // TODO: whether it matters at all.
+  // toTensorOp.setRestrict(true);
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp:98]: ditto.

File: lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+           << "linalg.generic op payload does not correspond with a "
+              "vectorizable contraction.";
+
+  // Insert reshape ops for output operand
+  auto res = scalarizeTensor(rewriter, loc, newOp.getResults()[0]);
+  rewriter.replaceOp(target, res);
+
+  results.push_back(newOp);
+
+  // return DiagnosedSilenceableFailure(target);
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp:297]: ditto.

File: lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+  auto shapeTy = cast<ShapedType>(opTy);
+
+  auto vecShape = shapeTy.getShape();
+  auto vecElemTy = cast<VectorType>(shapeTy.getElementType());
+  auto elemTy = vecElemTy.getElementType();
+  auto toMemRefVecTyOp = rewriter.create<bufferization::ToMemrefOp>(
+      loc, MemRefType::get(vecShape, vecElemTy), tensor);
+
+  //  SmallVector<int64_t> scalShape = llvm::to_vector(
+  //    llvm::concat<int64_t>(shapeTy.getShape(), vecElemTy.getShape()));
```
[the remark this answers is not in the document]
Comment (jsetoain) [reply lib/Dialect/AIEVec/TransformOps/AIEVecTransformOps.cpp:82]: This one fell through the cracks :-)

**Pull request merged**